PR upload validates comment file extension

adshmh opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

pullrequest upload now validates the extension of each comment file:
the only valid extensions are none and .json. This allows the upload to
fail if a .json comment file cannot be parsed, rather than posting all
the contents as the comment. Fixes #2168

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```release-note
Comment files under a PullRequest need to have either:
- no extension: for comment files that contain comment as plain text
- .json extension: for comment files in JSON format
```

dibyom commented:
/ok-to-test
cc @wlynch

sbwsg commented:
If we go ahead with this we'd need to document it somewhere; maybe in the pipeline resources doc?  Also bear in mind this binary is used by the [pullrequest task in the catalog](https://github.com/tektoncd/catalog/tree/v1beta1/pullrequest), so we'd need to update docs and tagged version there too.

I have some reservations about basing the behaviour on the file extension. But it might be fine in the context of the pullrequest resource and match user expectations?  I've never used it so I don't really have a strong idea.  @bobcatfish @wlynch does this seem like a good approach from your pov?  Basing the behaviour on the file extension?

/kind feature

wlynch commented:
The original purpose of this was so that tools could do something like `cat "asdf" > comment` to create comments without worrying about JSON validation, since the API only allows you to include the `body` field anyway. `.json` extensions were allowed so the same behavior could be used for updating comments.

I think the use case/concern given here is valid, though I think I'd prefer if we enable strict JSON validation on only `.json` extensions, and allow text for any other extension, this way users could also include things like `.md` extensions for markdown.
We should also probably make the extension check case insensitive(e.g. `.json`, `.JSON`, etc.) via strings.EqualFold.

We probably also want to control this behavior via a flag. This way existing users can have a mechanism for users to revert to the old behavior if this change breaks their workflow.

adshmh commented:
> If we go ahead with this we'd need to document it somewhere; maybe in the pipeline resources doc? Also bear in mind this binary is used by the [pullrequest task in the catalog](https://github.com/tektoncd/catalog/tree/v1beta1/pullrequest), so we'd need to update docs and tagged version there too.

Thank you for the review. I have updated the PR to include docs. I will also submit a PR against catalog.

adshmh commented:
> I think the use case/concern given here is valid, though I think I'd prefer if we enable strict JSON validation on only `.json` extensions, and allow text for any other extension, this way users could also include things like `.md` extensions for markdown.

Thank you for the review. I have updated the PR to address this comment.

> We should also probably make the extension check case insensitive(e.g. `.json`, `.JSON`, etc.) via strings.EqualFold.
> 
> We probably also want to control this behavior via a flag. This way existing users can have a mechanism for users to revert to the old behavior if this change breaks their workflow.

Thank you for the review. I have updated the PR and added a flag to allow overriding this behavior.

adshmh commented:
/retest

adshmh commented:
failures seem unrelated.
/retest

sbwsg commented on pkg/pullrequest/disk_test.go:
```diff
@@ -420,6 +420,131 @@ func readAndUnmarshal(t *testing.T, p string, v interface{}) {
+				t.Fatalf("Comments length does not match expected length. expected %d, got %d: %+v", len(tt.expectedComments), len(comments), comments)
+			}
+			for i, c := range tt.expectedComments {
+				if diff := cmp.Diff(comments[i], c); diff != "" {
```
I think the order of arguments is incorrect here; should be `(expected, received)`, which here would manifest as:

```go
if diff := cmp.Diff(c, comments[i]); diff != "" {
```

adshmh replied:
Thank you for the review. Fixed.

sbwsg commented on pkg/pullrequest/disk_test.go:
```diff
@@ -420,6 +420,131 @@ func readAndUnmarshal(t *testing.T, p string, v interface{}) {
+			}
+			for i, c := range tt.expectedComments {
+				if diff := cmp.Diff(comments[i], c); diff != "" {
+					t.Errorf("Get comments: -want +got: %s", diff)
```
This can be replaced with `diff.PrintWantGot(diff)`.  The `diff` package can be imported from `github.com/tektoncd/pipeline/test/diff`

adshmh replied:
Thank you for the review. Fixed.

adshmh commented:
> @adshmh I noticed one more fix to be made in a test but otherwise I think this PR is ready to go. Let me know if you're still interested in getting this merged. Will need a rebase and commits to be squashed as well then I'll lgtm and hopefully we can get this in ASAP!

Thank you for the review. Rebased and squashed commits.
